internal/mesh/internal/controllers/sidecarproxy/controller.go

```diff
@@ -230,6 +238,95 @@ func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req c
 	return nil
 }
 
+func (r *reconciler) workloadPortProtocolsFromService(
+	ctx context.Context,
+	fetcher *fetcher.Fetcher,
+	workload *types.DecodedWorkload,
+) (map[string]*pbcatalog.WorkloadPort, error) {
+
+	// Fetch all services for this workload.
+	serviceIDs := r.cache.ServicesForWorkload(workload.GetResource().GetId())
+
+	services := make([]*types.DecodedService, len(serviceIDs))
+
+	for i, serviceID := range serviceIDs {
+		svc, err := fetcher.FetchService(ctx, serviceID)
+		if err != nil {
+			return nil, err
+		}
+
+		// If service is not found, we should untrack it.
+		if svc == nil {
+			r.cache.UntrackService(serviceID)
+			continue
```

Comment:
You'll end your slice with `nil` this way when you skip. I think you need the 3-arg `make` and to use `append` instead of `[i]` to populate `services`

Reply:
great call. fixed!